src/rail_cmd.cpp

```diff
@@ -464,6 +465,25 @@ CommandCost CmdBuildSingleRail(TileIndex tile, DoCommandFlag flags, uint32 p1, u
 			if (ret.Succeeded()) ret = EnsureNoTrainOnTrack(tile, track);
 			if (ret.Failed()) return ret;
 
+			if (HasSignals(tile)) {
+				if (TracksOverlap(GetTrackBits(tile) | TrackToTrackBits(track))) {
+					/* If adding the new track causes any overlap, all signals must be removed first */
+					if (autoRemoveSignals) {
+						for (Track track_it = TRACK_BEGIN; track_it < TRACK_END; track_it++) {
+							if (HasTrack(tile, track_it) && HasSignalOnTrack(tile, track_it)) {
+								CommandCost ret_remove_signals = DoCommand(tile, track_it, 0, flags, CMD_REMOVE_SIGNALS);
+								if (ret_remove_signals.Failed()) return ret_remove_signals;
+								cost.AddCost(ret_remove_signals);
+							}
+						}
+					}
+					else
+					{
+						return_cmd_error(STR_ERROR_MUST_REMOVE_SIGNALS_FIRST);
```

Comment:
If you flip the `if()` around (so `if (!autoRemoveSignals) {`), you can return, and don't need the `else`. This makes the code more readable I think.